crates/iceberg-catalog/src/service/storage/gcs/mod.rs

```diff
@@ -0,0 +1,373 @@
+        if self.bucket != other.bucket {
+            return Err(UpdateError::ImmutableField("bucket".to_string()));
+        }
+
+        if self.key_prefix != other.key_prefix {
+            return Err(UpdateError::ImmutableField("key_prefix".to_string()));
+        }
+
+        Ok(())
+    }
+
+    #[must_use]
+    #[allow(clippy::unused_self)]
+    pub fn generate_catalog_config(&self, _: WarehouseIdent) -> CatalogConfig {
+        CatalogConfig {
+            defaults: HashMap::with_capacity(0),
+            overrides: HashMap::with_capacity(0),
+        }
+    }
+
+    /// Base Location for this storage profile.
+    ///
+    /// # Errors
+    /// Can fail for un-normalized profiles
+    pub fn base_location(&self) -> Result<Location, ValidationError> {
+        let prefix: Vec<String> = self
+            .key_prefix
+            .as_ref()
+            .map(|s| s.split('/').map(std::borrow::ToOwned::to_owned).collect())
+            .unwrap_or_default();
+        Location::from_str(&format!("gs://{}/", self.bucket))
+            .map(|mut l| {
+                l.extend(prefix.iter());
+                l
+            })
+            .map_err(|e| ValidationError::InvalidLocation {
+                reason: "Invalid GCS location.".to_string(),
+                location: format!("gs://{}/", self.bucket),
+                source: Some(e.into()),
+                storage_type: StorageType::Gcs,
+            })
+    }
+
+    /// Generate the table configuration for GCS.
+    pub(crate) async fn generate_table_config(
+        &self,
+        _: &DataAccess,
+        cred: Option<&GcsCredential>,
+        table_location: &Location,
+        storage_permissions: StoragePermissions,
+    ) -> Result<TableProperties, TableConfigError> {
+        let mut config = TableProperties::default();
+        if let Some(GcsCredential::ServiceAccountKey { key }) = cred {
+            let token = sts::downscope(
+                key,
+                &self.bucket,
+                table_location.clone(),
+                storage_permissions,
+            )
+            .await?;
+
+            config.insert(&gcs::Token(token.access_token));
+            config.insert(&gcs::ProjectId(key.project_id.clone()));
+
+            if let Some(expiry) = token.expires_in {
+                config.insert(&gcs::TokenExpiresAt(
+                    (std::time::SystemTime::now()
+                        .duration_since(std::time::UNIX_EPOCH)
+                        .unwrap()
+                        .as_secs()
+                        + expiry as u64)
+                        .to_string(),
+                ));
+            }
+        }
+
+        Ok(config)
+    }
+
+    fn normalize_key_prefix(&mut self) -> Result<(), ValidationError> {
+        if let Some(key_prefix) = self.key_prefix.as_mut() {
+            *key_prefix = key_prefix.trim_matches('/').to_string();
+            if key_prefix.starts_with(".well-known/acme-challenge/") {
```

Comment:
out of curiosity, why is this a problem?

Reply:
https://cloud.google.com/storage/docs/objects#naming

> Object names cannot start with .well-known/acme-challenge/.

There are also a few others, not sure if we should be checking them:

- Object names can contain any sequence of valid Unicode characters, of length 1-1024 bytes when UTF-8 encoded.
- Object names cannot contain [Carriage Return or Line Feed characters](https://en.wikipedia.org/wiki/Newline#Unicode).
- Objects cannot be named . or ...

Reply:
Hm, should be fine. Lets just accept an error in those cases